DSharpPlus/Clients/EventHandlerCollection.cs

```diff
@@ -9,7 +22,134 @@ namespace DSharpPlus;
 public sealed class EventHandlerCollection
 {
     /// <summary>
-    /// The delegate-based event handlers configured for this application.
+    /// Gets the registered handlers for a type, or an empty collection if none were configured. These are returned as
+    /// <see langword="object"/>, and individual elements must be casted to
+    /// <c>Func&lt;DiscordClient, TEventArgs, IServiceProvider, Task&gt;</c>.
+    /// </summary>
+    public IReadOnlyList<object> this[Type type]
+    {
+        get
+        {
+            return this.Handlers.TryGetValue(type, out IReadOnlyList<object>? handlers)
+                ? handlers
+                : ([]);
+        }
+    }
+
+    /// <summary>
+    /// The event handlers configured for this application.
+    /// </summary>
+    private FrozenDictionary<Type, IReadOnlyList<object>> Handlers
+        => this.immutableHandlers ?? AsImmutable();
+
+    private FrozenDictionary<Type, IReadOnlyList<object>>? immutableHandlers = null;
+
+    private readonly Dictionary<Type, List<object>> handlers = [];
+
+    /// <summary>
+    /// Registers a single simple event handler.
+    /// </summary>
+    /// <typeparam name="T">The type of event args this handler consumes.</typeparam>
+    /// <param name="handler">The event handler.</param>
+    public void Register<T>(Func<DiscordClient, T, Task> handler)
+        where T : DiscordEventArgs
+    {
+        if (this.handlers.TryGetValue(typeof(T), out List<object>? value))
+        {
+            value.Add(CanonicalizeSimpleDelegateHandler(Unsafe.As<SimpleHandler>(handler)));
+        }
+        else
+        {
+            List<object> temporary = [];
+            temporary.Add(CanonicalizeSimpleDelegateHandler(Unsafe.As<SimpleHandler>(handler)));
+
+            this.handlers.Add(typeof(T), temporary);
+        }
+    }
+
+    /// <summary>
+    /// Registers all type-wise event handlers implemented by a specific type.
     /// </summary>
-    public ConcurrentDictionary<Type, ConcurrentBag<Delegate>> DelegateHandlers { get; } = [];
+    /// <typeparam name="T">The type whose handlers to register.</typeparam>
+    public void Register<T>()
+        where T : IEventHandler
+    {
+        foreach ((Type args, Handler handler) in CanonicalizeTypeHandlerImplementations(typeof(T)))
+        {
+            if (this.handlers.TryGetValue(args, out List<object>? value))
+            {
+                value.Add(handler);
+            }
+            else
+            {
+                List<object> temporary = [];
+                temporary.Add(handler);
```

Comment:
If you're just calling .Add anyway, why not use the collection expression syntax to make the list? Clarity, perhaps? 

For context I mean `this.handlers.Add(args, [handler]);`

Reply:
because the list needs to be `List<object>` specifically. it can be `List<object> temporary = [handler];` though